### PR TITLE
E(30)exch-ind in SAPT2+3 without the S^2 approximation

### PR DESCRIPTION
**Pull request by konpat**
## Description
Adds an initial implementation of the E(30)exch-ind SAPT correction without the single exchange (S^2) approximation, introduced in https://doi.org/10.1063/5.0035050. Original Psi4NumPy implementation by Jonathan Waldrop (@jwaldrop107), ported to C++ by Konrad Patkowski. 

The resulting numbers for a bunch of tests match the SI of the above paper and the calculation takes a tiny fraction of the total SAPT2+3 time. E(30)exch-ind,resp is approximated by scaling unrelaxed E(30)exch-ind in the same way as before and is also printed.

## Usage
The nonapproximated E(30)exch-ind and E(30)exch-ind,resp numbers will be calculated by `energy('sapt2+3')` if `do_ind30_exch_sinf` is set to `True`. 

## Technical Questions and Unresolved Issues
- [x] There's certainly room for code cleanup.
- [x] I think `libsapt_solver` is the most logical home for this code and `energy('sapt2+3')` its most logical invocation - after all, this correction forms a part of SAPT2+3 only, not the lower SAPT levels. The E(30)ind correction needs to be calculated beforehand, both for subtracting its value from the total (ind+ex-ind) calculated here, and for generating the third-order induction amplitudes that are reused here. However, unlike most of the stuff in `libsapt_solver`, the code is AO-based and the two-electron integrals enter only via generalized J/K matrices. Therefore, I had to export some additional AO quantities in `libsapt_solver/sapt.h` and I don't know if I did this in an optimal way.
- [x] Note to self: frozen core still needs to be checked.

## Design (Big-Picture) Questions - input from @CDSherrill @dgasmith @loriab and others will be highly beneficial!
- [ ] The three new(ish) beyond-S^2 corrections, E(20)exch-ind, E(20)exch-disp, and E(30)exch-ind cannot be currently calculated by a single SAPT call: the first two require `energy('sapt(dft)')` and the last one `energy('sapt2+3')`. That's quite unfortunate and we need a better workflow, possibly by making the non-S^2 E(20)exch-ind,resp and E(20)exch-disp available as part of `energy('sapt0')`.
- [ ] What should go into aggregate SAPT energy values, S^2 or non-S^2? I made the code pick the non-S^2 value of E(30)exch-ind,resp if available, but I'd be happy to revert it back to picking E(30)exch-ind,resp(S^2). At present, I don't think the non-S^2 values of E(20)exch-ind,resp and E(20)exch-disp are ever included in total SAPT numbers, maybe they should?
- [ ] Finally, the option names such as `do_ind30_exch_sinf` are somewhat unfortunate because they propagate the misconception that you include effects all the way to S^infinity. In reality, the largest power of overlap integral S you ever need is min(NelA,NelB) - after that, you run out of electrons to exchange between A and B. **Nonapproximated**, **Nonexpanded**, and **beyond-S^2** are all formally more correct names than S^infinity.

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [ ] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:611]: Please use `Matrix::vector_dot` instead of manual BLAS calls, for readability.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:560]: Please use `Matrix::vector_dot` instead of manual BLAS calls, for readability.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:375]: The `libmints` function gets block allows you to get a block of a Matrix. You _could_ use this to get the four blocks of the D^-1 matrix as their own matrices, and then 8 of the "New stuff" DGEMM calls become simple doublets. I expect your half-transforms and your xi-dependent terms simplify as well. There will be some additional copy overhead from this approach, but the code becomes cleaner. I won't insist on you doing this, but I do recommend you consider it.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:462]: Recurring issue: I don't understand why you clone a matrix, zero it, and then add the matrix you just cloned. Just clone it.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:569]: I believe you can just hit these with doublet?

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:615]: Q for Lori: Do we want to define the global? I know we're shifting to wavefunction variables.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:553]: These are good candidates for `auto`.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:430]: `auto&` candidates. `const auto&` below.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:34]: Why is this header needed?

**Review by JonathonMisiewicz — COMMENTED** [review]
Thanks for the feature. I'm not a SAPT expert, so I'll leave your big picture questions to Lori.

Overarching comment: New features always need documentation and tests. This currently has none. Do you need any guidance on what that looks like?

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc]: It wasn't. I copied and pasted too much.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc]: Changed.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc]: Changed.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc]: Yes, I can, and now I do.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc]: My bad. Fixed throughout.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc:374]: I see your point, it's just that I'm quite confused about the syntax of the slice definition for `matrix->get_block`.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc]: Done!

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (loriab) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc:590]: It's a lot simpler to switch over from P::e.globals to wfn module-by-module (or at least method-by-method). Looks like over libsapt_solver are still globals, so this new one is good as global.

**Comment by konpat** [comment]
Thanks much for the review and helpful corrections!

Yes, I am fully aware that adding tests and documentation is necessary, I think I know how to go about doing this, and I will certainly add these before the PR is merge-ready. 

It would be really nice to have a test that checks all three beyond-S^2 corrections: E(20)exch-ind, E(20)exch-disp, and E(30)exch-ind. It would require two SAPT calculations at the moment though.

**Comment by konpat** [comment]
Yes, @loriab, I'm looking forward to discussing the naming of SAPT variants/calls, and Psicon will be a good opportunity.

I very much like the name `do_ind30_exch_full`.

**Comment by JonathonMisiewicz** [comment]
I'm marking this PR as blocked until PsiCon, pending Auburn/GATech discussions.

**Comment by konpat** [comment]
This PR is ready to be unblocked and reviewed. A test has been added. Frozen-core works as expected, that is, the number is unchanged (just like regular E(30)exch-ind). The numbers agree with the original paper. @jwaldrop107 - you might want to take a look and check if everything's still working in your hands.

We had a very productive discussion with @CDSherrill @loriab @zachglick last week. We have some ideas how to make the SAPT workflows, including the stuff without the S^2 approximation, simpler and more consistent. We are also thinking of renaming all **Sinf** to **Sall**, but that's a story for a separate PR.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:338]: Unused variable.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:346]: I'd prefer `Matrix` over `shared_ptr<Matrix>`, unless there's a reason we need the pointer.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:351]:
```suggestion
    auto Sab = linalg::triplet(CoccA_, Smat_, CoccB_, true, false, false);
```

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:354]: This can also just be a regular `Matrix`.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:377]:
```suggestion
    auto Sbr = linalg::triplet(CoccB_, Smat_, CvirA_, true, false, false);
    auto Sas = linalg::triplet(CoccA_, Smat_, CvirB_, true, false, false);
```

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:432]: This is another case to use `auto`.

**Review by JonathonMisiewicz — COMMENTED** [review]
I have some very nitpicky suggestions, but this code looks pretty good to me.

Do you think the documentation needs to be updated?

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc]: Done.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc]: Removed.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc:345]: Then I need to reveal my ignorance and ask a basic question: how to do this so that `uAR->load()` and `uAR->vector_dot()` still work?

Comment (JonathonMisiewicz) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc:345]: Just change `->` to `.`.

`->` is the syntax for "this object is not a Matrix but a memory address of a Matrix, so find the Matrix at this memory address and then call this Matrix method". `.` is the syntax for "this object is a Matrix, so call this Matrix method".

Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc:345]: OK, done, thanks!

**Review by JonathonMisiewicz — APPROVED** [review]
The code looks good to me - thanks for the new feature!

I'll leave it to the SAPT experts to decide if this needs documentation on the main SAPT page beyond whatever gets auto-added when a new option is introduced.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (jeffschriber) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:339]: I don't think this threading variable is needed

**Review by jeffschriber — COMMENTED** [review]
This is great! Code looks good to me, and I think the addition to the manual is good and necessary.

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (konpat) [reply psi4/src/psi4/libsapt_solver/exch-ind30.cc]: It's not. I removed it.

**Review by jeffschriber — APPROVED** [review]

**Comment by konpat** [comment]
From my perspective, everything is in good shape now, including the documentation.

**Review by loriab — APPROVED** [review]
lgtm, thank you!

The issue of totaling qcvars is handled in python and so isn't covered here. But that can be addressed together with some of the other changes from the Dec sapt mini conf in a future PR.

Will merge this week unless any other comments come up!

File: psi4/src/psi4/libsapt_solver/exch-ind30.cc
Comment (loriab) [inline psi4/src/psi4/libsapt_solver/exch-ind30.cc:356]: There's also a `Matrix.general_inverse` that may be of interest: https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/matrix.h#L959-L992

**Pull request merged**
